README.markdown
lockrun - Run cron job with overrun protection
==============================================

When doing network monitoring, it's common to run a cron job every five
minutes (the standard interval) to roam around the network gathering
data. Smaller installations may have no trouble running within this
limit, but in larger networks or those where devices are often
unreachable, running past the five-minute mark could happen frequently.

The effect of running over depends on the nature of the monitoring
application: it could be of no consequence or it could be catastrophic.
What's in common is that running two jobs at once (the old one which ran
over, plus the new one) slows down the new one, increasing the risk that
*it* will run long as well.

This is commonly a cascading failure which can take many polling
sessions to right itself, which may include lost data in the interim.

Our response has been to create this tool, `lockrun`, which serves as a
protective wrapper. Before launching the given command, it insures that
another instance of the same command is not already running.

Build and Install
=================

This tool is published in the form of portable C source code, and it can
be compiled on any Linux/UNIX platform that provides a development
environment.

To build and install it from the command line:

    $ gcc lockrun.c -o lockrun
    $ sudo cp lockrun /usr/local/bin/

Now we'll find `lockrun` in the usual place: `/usr/local/bin/`.

We'll note that though portable, this program is nevertheless designed
only to run on UNIX or Linux systems: it certainly won't build and run
properly on a Windows computer.

Furthermore, file locking has always been one of the more problemtic
areas of portability, there being several mechanisms in place. `lockrun`
uses the `flock()` system call, and this of course requires low-level OS
support.

We've tested this in FreeBSD and Linux, but other operating systems
might trip over compilation issues. We welcome portability reports (for
good or bad).

We've also received a report that this works on Apple's OS X.

Example Usage
=============

Once `lockrun` has been built and installed, it's time to put it
to work. This is virtually always used in a crontab entry, and the
command line should include the name of the lockfile to use as well
as the command to run.

This entry in a crontab file runs the Cacti poller script every five
minutes, protected by a lockfile:

    */5 * * * * /usr/local/bin/lockrun --lockfile=/tmp/cacti.lockrun -- /usr/local/bin/cron-cacti-poller

The file used, `/tmp/cacti.lockrun`, is created (if necessary), the lock
acquired, and closed when finished. At no time does `lockrun` perform
any file I/O: the file exists only to be the subject of locking
requests.

Note that everything up to the standalone `--` is considered an option
to `lockrun`, but everything after is the literal command to run.

The example provided here is a run-or-nothing instance: if the lock
cannot be acquired, the program exits with a failure message to the
standard error stream, which hopefully is routed back to the user via an
email notification:

`ERROR: cannot launch [command line] - run is locked`

This mechanism effectively *skips a polling run*, but this may be the
only option when polling runs long periodically. If one polling run goes
quite long, it's conceivable that multiple subsequent jobs could be
stacked behind the slow one, and never getting caught up.

But if most jobs complete very rapidly, adding the `--wait`
parameter might allow the system to catch up after a lone straggler
runs long.


However one organizes this, one can't avoid being concerned with runs
which are locked often. An inability to complete a polling run on time
indicates a resource-allocation problem which is not actually fixed by
skipping some data.

If this happens regularly, it's important to track down what's causing
the overruns: lack of memory? inadequate CPU? serialized jobs which
could benefit from parallelization or asynchronous processing?

There is no substitute for actual human observation of important
systems, and though `lockrun` may forestall a monitoring meltdown, it
doesn't replace paying attention. It is **not** an advanced command
queuing system.

Locking Behavior
================

We've been asked why we do this in a C program and not a simple shell
script: the answer is that we require bulletproof, no-maintenance
protection, and that's very hard to do with shell scripting.

With touch-a-file locking, there's a chance that the lockfile can be
left around after everything is done: what if the cron job has run long,
and the administrator killed everything associated with the job? What
about a system crash leaving the lockfile around? What if there's a
fatal error in `lockrun` itself? All of these leave the lockfile around
in the system for the next run to trip over.

One could make this mechanism smarter by including the `PID` of the
locking process inside the file, and then using `kill(*pid*,0)` to see
if that process exists, but PIDs are reused, and it's possible to have a
false positive (i.e., when the previous `lockrun` has finished, but some
*other* process has taken that PID slot). We've always disliked the
nondeterminism of this mechanism.

So we required a mechanism which provided guaranteed, bulletproof
cleanup at program exit, and no chance of false positives. Though one
can find numerous mechanisms for this, use of file locks is the easiest
to code and understand. Setting a lock automatically tests for the
previous lock, and this means no race conditions to worry about. When
the file is closed, locks evaporate.

Note that file locking under UNIX is typically *advisory* only: A lock
placed by one process is only honored by other processes who chose to
check the lock first. Any process with suitable permissions is free to
read or write anything without regard to locks.

Advisory locking works on the honor system, but they're entirely
appropriate for our use here.

Finally, we'll note that our locking mechanism is only designed to
prevent two lock-protected processes from running at once; It is *not* a
queuing system.

When using the `--wait` parameter, it's entirely possible to have many
processes stacked up in line behind a prior long-running process. When
the long-running process exists, it's impossible to predict which of the
waiting processes will run next, and it's probably not going to be done
in the order in which they were launched. Users with more sophisticated
queuing requirements probably need to find a different mechanism.

Command-Line Options
====================

`lockrun` supports GNU-style command-line options, and this includes
using `--` to mark their end:

    $ lockrun [options] -- [command]

The actual command after `--` can have any arguments it likes, and they
are entirely uninterpreted by `lockrun`.

We'll note that command-line redirection (`> /dev/null`, etc.) is not
supported by this or the command which follows -- it's handled **by the
calling shell**. This is the case whether it's run from cron or not.

 * `--idempotent`

  > Allows silent successful exit when lock contention is encountered.

 * `--lockfile=[filename]`

  > Specify the name of a file which is used for locking. This filename
  > is created if necessary (with mode 0666), and no I/O of any kind is
  > done. This file is never removed.

 * `--maxtime=[N]`

  > The script being controlled ought to run for no more than *N*
  > seconds, and if it's beyond that time, we should report it to the
  > standard error stream (which probably gets routed to the user via
  > cron's email).


 * `--wait`

  > When a pre-existing lock is found, this program normally exits with
  > error, but adding the `--wait` parameter causes it to loop, waiting
  > for the prior lock to be released.

 * `--verbose`

  > Show a bit more runtime debugging.

  * `--`

  > Mark the end of the options, the actual command to run follows.

History
=======

 * 2010/10/04 - added idempotency to allow run lock contention to be treated as a no-op (Mike Cerna, Groupon)
 * 2009/06/25 — added lockf() support for Solaris 10 (thanks to Michal Bella)
 * 2009/03/09 — Tracked on GitHub by Peter Harkins.
 * 2006/06/03 — initial release by Stephen J. Friedl. <http://unixwiz.net/archives/2006/06/new_tool_lockru.html>

License
=======

This software was places in the public domain by Stephen J. Friedl.
All contributions are also in the public domain.

The original license reads:

> Written by : Stephen J. Friedl  
>              Software Consultant  
>              Southern California USA  
>              steve@unixwiz.net  
>              http://www.unixwiz.net/tools/  
>   
>      ===================================================================  
>      ======== This software is in the public domain, and can be ========  
>      ======== used by anybody for any purpose                   ========  
>      ===================================================================  


debian/changelog
lockrun (2:1.1.2-2) unstable; urgency=low

  [ Allard Hoeve ]
  * Set RC when execvp fails

  [ Cipri ]
  * Allow uppercase -Q, as before
  * Update man page to consistently use -Q, eventhough we accept -q as well

  [ Peter Harkins ]
  * Credit contributors

  [ Allard Hoeve ]
  * Add manpage
  * Add Makefile
  * Restructure README so that a manpage can be generated
  * Add ruby-ronn to Depends for manpages

 -- Allard Hoeve <allard@byte.nl>  Mon, 12 Aug 2013 14:39:48 +0200

lockrun (2:1.1.2) unstable; urgency=low

  * Support -Q as well as -q (Cipri).

 -- Allard Hoeve <allard@byte.nl>  Wed, 24 Jul 2013 15:09:08 +0200

lockrun (1.1.2) unstable; urgency=low

  * New upstream version.
    - new option --retries to be used with --wait
  * Merge with github version at https://github.com/pushcx/lockrun
    - new option --idempotent is an alias to --quiet
    - exit using command's exitcode

 -- Allard Hoeve <allard@byte.nl>  Wed, 24 Jul 2013 14:07:13 +0200

lockrun (1.1.1) unstable; urgency=low

  * New upstream version.
    - new option --quiet
    - support flock on platforms that have it

 -- Gertjan Oude Lohuis <gertjan@byte.nl>  Tue, 04 Dec 2012 11:39:21 +0100

lockrun (1.0.3) unstable; urgency=low

  * Initial Release.

 -- Roy van Baekel <roy@byte.nl>  Wed, 12 Jan 2011 12:53:45 +0100


lockrun.c
/*
 * See README.markdown for build, install, and usage instructions.
 */

#include <stdio.h>
#include <stdarg.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/file.h>

static const char Version[] = "lockrun 1.1.2 - 2013-04-26 - steve@unixwiz.net";

#ifndef __GNUC__
# define __attribute__(x)	/* nothing */
#endif


#define STRMATCH(a,b)		(strcmp((a),(b)) == 0)

#define UNUSED_PARAMETER(v)	((void)(v))

#define	TRUE	1
#define	FALSE	0

static const char	*lockfile = 0;
static int		wait_for_lock = FALSE;
static mode_t		openmode = 0666;
static int		sleeptime = 10;		/* seconds */
static int		retries = 0;
static int		Verbose = FALSE;
static int		Maxtime  = 0;
static int		Quiet = FALSE;

static char *getarg(char *opt, char ***pargv);

static void die(const char *format, ...)
		__attribute__((noreturn))
		__attribute__((format(printf, 1, 2)));

#ifdef F_TLOCK
# define WAIT_AND_LOCK(fd) lockf(fd, F_TLOCK,0)
#else
# define WAIT_AND_LOCK(fd) flock(fd, LOCK_EX | LOCK_NB)
#endif

static const char *const helptext[] = {
	"Usage: lockrun [options] -- command args...",
	"",
	"  --help        Show this brief help listing",
	"  --version     Report the version info",
	"  --            Mark the end of lockrun options; command follows",
	"  --verbose     Show a bit more runtime debugging",
	"    -V",
	"  --idempotent  See --quiet",
	"    -I",
	"  --quiet       Exit quietly (and with success) if locked",
	"    -q",
	"    -Q",
	"  --lockfile=F  Specify lockfile as file <F>",
	"    -L=F",
	"  --wait        Wait for lockfile to appear (else exit on lock)",
	"    -W",
	"",
	" Options with --wait:",
	"",
	"  --sleep=T     Sleep for <T> seconds on each wait loop",
	"    -S=T",
	"  --retries=N   Attempt <N> retries in each wait loop",
	"    -R=N",
	"  --maxtime=T   Wait for at most <T> seconds for a lock, then exit",
	0,
};

static void show_help(const char *const *);

int main(int argc, char **argv)
{
	char	*Argv0 = *argv;
	int	rc;
	int	lfd;
	pid_t	childpid;
	time_t	starttime;
	int	attempts = 0;

	UNUSED_PARAMETER(argc);

	time(&starttime);

	for ( argv++ ; *argv ; argv++ )
	{
		char    *arg = *argv;
		char	*opt = strchr(arg, '=');

		/* the -- token marks the end of the list */

		if ( strcmp(*argv, "--") == 0 )
		{
			argv++;
			break;
		}

		if (opt) *opt++ = '\0'; /* pick off the =VALUE part */

		if ( STRMATCH(arg, "--version") )
		{
			puts(Version);
			exit(EXIT_SUCCESS);
		}

		else if ( STRMATCH(arg, "--help"))
		{
			show_help(helptext);
			exit(EXIT_SUCCESS);
		}

		else if ( STRMATCH(arg, "-L") || STRMATCH(arg, "--lockfile"))
		{
			lockfile = getarg(opt, &argv);
		}

		else if ( STRMATCH(arg, "-W") || STRMATCH(arg, "--wait"))
		{
			wait_for_lock = TRUE;
		}

		else if ( STRMATCH(arg, "-S") || STRMATCH(arg, "--sleep"))
		{
			sleeptime = atoi(getarg(opt, &argv));
		}

		else if ( STRMATCH(arg, "-R") || STRMATCH(arg, "--retries"))
		{
			retries = atoi(getarg(opt, &argv));
		}

		else if ( STRMATCH(arg, "-T") || STRMATCH(arg, "--maxtime"))
		{
			Maxtime = atoi(getarg(opt, &argv));
		}

		else if ( STRMATCH(arg, "-V") || STRMATCH(arg, "--verbose"))
		{
			Verbose++;
		}

		else if ( STRMATCH(arg, "-Q") || STRMATCH(arg, "-q") || STRMATCH(arg, "--quiet") || STRMATCH(arg, "-I") || STRMATCH(arg, "--idempotent"))
		{
			Quiet = TRUE;
		}

		else
		{
			die("ERROR: \"%s\" is an invalid cmdline param", arg);
		}
	}

	/*----------------------------------------------------------------
	 * SANITY CHECKING
	 *
	 * Make sure that we have all the parameters we require
	 */
	if (*argv == 0)
		die("ERROR: missing command to %s (must follow \"--\" marker) ", Argv0);

	if (lockfile == 0)
		die("ERROR: missing --lockfile=F parameter");

	/*----------------------------------------------------------------
	 * Open or create the lockfile, then try to acquire the lock. If
	 * the lock is acquired immediately (==0), then we're done, but
	 * if the lock is not available, we have to wait for it.
	 *
	 * We can either loop trying for the lock (for --wait), or exit
	 * with error.
	 */

	if ( (lfd = open(lockfile, O_RDWR|O_CREAT, openmode)) < 0)
		die("ERROR: cannot open(%s) [err=%s]", lockfile, strerror(errno));

	while ( WAIT_AND_LOCK(lfd) != 0 )
	{
		attempts++;

		if ( ! wait_for_lock )
		{
			if ( Quiet)
				exit(EXIT_SUCCESS);
			else
				die("ERROR: cannot launch %s - run is locked", argv[0]);
		}

		if ( retries > 0 && attempts >= retries )
		{
			die("ERROR: cannot launch %s - run is locked (after %d attempts)", argv[0], attempts);
		}

		/* waiting */
 		if ( Verbose ) printf("(locked: sleeping %d secs, after attempt #%d)\n", sleeptime, attempts);

		sleep(sleeptime);
	}

	fflush(stdout);

	/* run the child */


	if ( (childpid = fork()) == 0 )
	{
		/* IN THE CHILD */

		close(lfd);		/* don't need the lock file */

		/* Make ourselves a process group leader so that this and all
		 * child processes can be sent a signal as a group. This may
		 * be used someday to support a --kill option, though this is
		 * is still tricky.
		 *
		 * PORTING NOTE: if "setsid" is undefined on your platform,
		 * just comment it out and send me an email with info about
		 * the build environment.
		 */
		(void) setsid();

		/* Set rc to the result of execvp. This lets the parent know we failed. */
		rc = execvp(argv[0], argv);
	}
	else if ( childpid > 0 )
	{
		/* IN THE PARENT */

		time_t endtime;
		pid_t  pid;
		int    status;

		if ( Verbose )
		    printf("Waiting for process %ld\n", (long) childpid);

		pid = waitpid(childpid, &status, 0);
		rc = WEXITSTATUS(status);

		time(&endtime);

		endtime -= starttime;

		if ( Verbose || (Maxtime > 0  &&  endtime > Maxtime) )
		    printf("pid %d exited with status %d, exit code: %d (time=%ld sec)\n",
			   pid, status, rc, endtime);
	}
	else
	{
		die("ERROR: cannot fork [%s]", strerror(errno));
	}

	exit(rc);
}


/*! \fn static char *getarg(char *opt, char ***pargv)
 *  \brief A function to parse calling parameters
 *
 *	This is a helper for the main arg-processing loop: we work with
 *	options which are either of the form "-X=FOO" or "-X FOO"; we
 *	want an easy way to handle either one.
 *
 *	The idea is that if the parameter has an = sign, we use the rest
 *	of that same argv[X] string, otherwise we have to get the *next*
 *	argv[X] string. But it's an error if an option-requiring param
 *	is at the end of the list with no argument to follow.
 *
 *	The option name could be of the form "-C" or "--conf", but we
 *	grab it from the existing argv[] so we can report it well.
 *
 * \return character pointer to the argument
 *
 */
static char *getarg(char *opt, char ***pargv)
{
	const char *const optname = **pargv;

	/* option already set? */
	if (opt) return opt;

	/* advance to next argv[] and try that one */
	if ((opt = *++(*pargv)) == 0)
		die("ERROR: option %s requires a parameter", optname);

	return opt;
}

/*
 * die()
 *
 *	Given a printf-style argument list, format it to the standard error,
 *	append a newline, then exit with error status.
 */

static void die(const char *format, ...)
{
va_list	args;

	va_start(args, format);
	vfprintf(stderr, format, args);
	putc('\n', stderr);
	va_end(args);

	exit(EXIT_FAILURE);
}

static void show_help(const char * const *p)
{
	puts(Version);
	puts("");

	for ( ; *p; p++ )
	{
		puts(*p);
	}
}
